### ISSUE TITLE
Feature: Canvas: Reuse evaluated signatures

### ISSUE THREAD

**Issue by davemt** [celery/celery#893]
Consider the following (where A, B, C, D, E are tasks):

```
sub_A = A.s()
chord_D = chord([sub_A, B.s()], D.s())
chord_E = chord([sub_A, C.s()], E.s())
res = group(chord_D, chord_E)()
```

This creates two separate invocations for task A.  For many cases, we would want the subtask for A to be invoked just once, and the work shared.

It seems that manually giving the subtask a uuid leads to the desired outcome, but it would be nice to have support for this in the API.

**Comment by hevy-CC4477** [comment]
I am not sure to get it. Does sub_A will be executed twice in this case or not?

```python
sub_A = A.s().freeze()
chord_D = chord([sub_A, B.s()], D.s())
chord_E = chord([sub_A, C.s()], E.s())
```

**Comment by ask** [comment]
Yes!

**Comment by neoascetic** [comment]
Want this :)

**Comment by GlennS** [comment]
Just a note that this appears to be the same issue as [discussed here](https://groups.google.com/forum/#!topic/celery-users/EKQRkeVvGno).

I wasn't able to figure out how to pass a list of task_ids to the underlying chord implementation.

**Comment by ask** [comment]
Closing this, as with current resources it will take 10 years to complete.

**Comment by auvipy** [comment]
https://github.com/celery/celery/issues/3198

**Comment by alanjds** [comment]
Can the more involved people tell what should be done to fix this and #4562 ?

Having a roadmap could allow other people w/ more free hands to take over and take not years to complete ;)

**Comment by auvipy** [comment]
it's on the 5.0 roadmap.

**Comment by auvipy** [comment]
I think we should move this to next milestone

**Comment by thedrow** [comment]
I agree. This can be done only with an event loop as we have to check the backend first for a result which means waiting.
This requires an event loop.

In addition, it requires the task to be declared as idempotent which is something we currently don't support.

**Comment by jpsnyder** [comment]
I just ran into this issue myself. Really putting a damper on my excitement to use celery.

I was wondering if anybody knows a good workaround (without multiple blocking stages)?

Basically we need a "reverse chord", where instead of the results of a group of tasks feeding into a single task, we have the result of a single task feeding into a group of many tasks.

```python
celery.reverse_chord(add.s(1, 2), [sub.s(3), sub.s(4), sub.s(5)]) 
```

**Comment by sjkoelle** [comment]
Has there been any progress?  Support for arbitrary DAGs would be fantastic.

**Comment by PythonFZ** [comment]
Is this still on the Roadmap? I do have my project workflow as `networkx` graph and I am trying to run it using celery phasing exactly this issue.

**Comment by jpsnyder** [comment]
It's been a long time since I used celery, but I vaguely remember trying that and finding out it doesn't work. 

I ended up abandoning Celery and moved to RQ. There I was able to create a custom decorator that causes the task to wait for the main task to finish before starting.
